Add OPENING & CLOSING state to MySensors cover

FidgetyRat opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Added support for OPENING and CLOSING states to cover devices using a combination of the required mysensor V_ variables according to the existing documentation.

Simplified the determination of the cover's state by use of a new enumeration and single method allowing the state to be used by all three HomeAssistant query methods.

Minor change to devices.py to allow V_UP, V_DOWN, and V_STOP to be converted to STATE_ON and STATE_OFF.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [X] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [X] The code change is tested and works locally.
- [ ] Local tests pass. **Your PR cannot be merged unless tests pass**
- [X] There is no commented out code in this PR.
- [ ] I have followed the [development checklist][dev-checklist]
- [ ] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

FidgetyRat committed:
Added OPENING & CLOSING State Support

Added support for OPENING and CLOSING states using a combination of
the required V_ variables. Simplified the determination of the
cover's state by use of a new enumeration and single method allowing
the state to be used by all three HomeAssistant query methods.

homeassistant commented:
Hello @FidgetyRat,

When attempting to inspect the commits of your pull request for CLA signature status among all authors we encountered commit(s) which were not linked to a GitHub account, thus not allowing us to determine their status(es).

The commits that are missing a linked GitHub account are the following:

- [`30faa670d70399db531e7498787bb1d031351415`](https:&#x2F;&#x2F;github.com&#x2F;home-assistant&#x2F;core&#x2F;pull&#x2F;47285&#x2F;commits&#x2F;30faa670d70399db531e7498787bb1d031351415) - This commit has something that looks like an email address (rderelanko@gmail.com). Maybe try linking that to GitHub?.

Unfortunately, **we are unable to accept this pull request until this situation is corrected.**

Here are your options:

1. If you had an email address set for the commit that simply wasn't linked to your GitHub account you can link that email now and it will retroactively apply to your commits. The simplest way to do this is to click the link to one of the above commits and look for a blue question mark in a blue circle in the top left. Hovering over that bubble will show you what email address you used. Clicking on that button will take you to your email address settings on GitHub. Just add the email address on that page and you're all set. GitHub has more information about this option [in their help center](https://help.github.com/articles/setting-your-email-in-git/#commits-on-github-arent-linking-to-my-account).

2. If you didn't use an email address at all, it was an invalid email, or it's one you can't link to your GitHub, you will need to change the authorship information of the commit and your global Git settings so this doesn't happen again going forward. GitHub provides some great instructions on how to change your authorship information [in their help center](https://help.github.com/articles/setting-your-email-in-git/).
   * If you only made a single commit you should be able to run
      ```
      git commit --amend --author="Author Name <email@address.com>"
      ```
      (substituting Author Name and email@address.com for your actual information) to set the authorship information.
   * If you made more than one commit and the commit with the missing authorship information is not the most recent one you have two options:
       1. You can re-create all commits missing authorship information. This is going to be the easiest solution for developers that aren't extremely confident in their Git and command line skills.
       2. You can use [this script](https://help.github.com/articles/changing-author-info/) that GitHub provides to rewrite history. **Please note:** this should be used only if you are very confident in your abilities and understand its impacts.
   * Whichever method you choose, I will come by to re-check the pull request once you push the fixes to this branch.

We apologize for this inconvenience, especially since it usually bites new contributors to Home Assistant. We hope you understand the need for us to protect ourselves and the great community we all have built legally. The best thing to come out of this is that you only need to fix this once and it benefits the entire Home Assistant and GitHub community.

Thanks, I look forward to checking this PR again soon! :heart:

homeassistant commented:
Hi @FidgetyRat, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;core#47285).

Once you do that we will be able to review and accept this pull request.

Thanks!

FidgetyRat committed:
Fixes for HomeAssistant Style

Corrections to style to allow flake8, isort, and black to pass.

MartinHjelmare commented on homeassistant/components/mysensors/cover.py:
```diff
@@ -14,6 +15,15 @@
+
+    OPEN = 0
+    OPENING = 1
+    CLOSED = 2
+    CLOSING = 3
```
Let's make the sequence go from open to closed.
```suggestion
    CLOSING = 2
    CLOSED = 3
```

FidgetyRat replied:
Makes sense to me. You never know if someone will want to use the ordinals for something.

MartinHjelmare commented on homeassistant/components/mysensors/cover.py:
```diff
@@ -43,13 +53,44 @@ async def async_discover(discovery_info):
+        else:
+            amount = 100 if self._values.get(set_req.V_LIGHT) == STATE_ON else 0
+
+        if v_up and (not v_down) and (not v_stop):
+            return CoverState.OPENING
+        if (not v_up) and v_down and (not v_stop):
+            return CoverState.CLOSING
+        if (not v_up) and (not v_down) and v_stop and amount == 0:
```
Looks like we can drop the parentheses:
https://docs.python.org/3/reference/expressions.html#operator-precedence
```suggestion
        if v_up and not v_down and not v_stop:
            return CoverState.OPENING
        if not v_up and v_down and not v_stop:
            return CoverState.CLOSING
        if not v_up and not v_down and v_stop and amount == 0:
```

FidgetyRat replied:
Why not. I'll add to the next batch.

MartinHjelmare commented on homeassistant/components/mysensors/cover.py:
```diff
@@ -14,6 +15,15 @@
+class CoverState(Enum):
+    """An enumeration of the standard cover states."""
+
+    OPEN = 0
+    OPENING = 1
+    CLOSED = 2
+    CLOSING = 3
```
We could also just make these strings?
```suggestion
    OPEN = "open"
    OPENING = "opening"
    CLOSING = "closing"
    CLOSED = "closed"
```

FidgetyRat replied:
Do you have a specific reason for it? I personally prefer to keep enumerations numeric as you can access the enumeration name via Enum.name (at least in terms of the capitalized formal name). I have no issue making the change if you have a use case you can see that would use it.

That said, we probably should have @unique applied to the enum either way.

MartinHjelmare replied:
I guess I'm wondering what's the use case for the numbers. There's nothing saying they should be numbers. It's a local definition that we use just here and we don't use the numbers.

FidgetyRat replied:
Numeric ordinals are just the default value. The Enum has to be bound to some constant value, and you are correct, you can use Strings for this local use, it doesn't exactly matter. I tend to prefer sequential integers as they can be used for something like an index to a list of other constants, but as you said, I'm not doing anything with them in this local case.

MartinHjelmare replied:
I'm ok with keeping them as numbers.

MartinHjelmare reviewed (commented):
Thanks!

FidgetyRat commented on homeassistant/components/mysensors/cover.py:
```diff
@@ -1,4 +1,5 @@
 """Support for MySensors covers."""
+from enum import Enum
```
```suggestion
from enum import Enum, unique
```

FidgetyRat commented on homeassistant/components/mysensors/cover.py:
```diff
@@ -14,6 +15,15 @@
 _LOGGER = logging.getLogger(__name__)
 
 
+class CoverState(Enum):
```
```suggestion
@unique
class CoverState(Enum):
```

FidgetyRat committed:
Peer Review Changes

Added @unique to the main enumeration. Removed unnecessary parens from
door state logic. Reordered CLOSING and CLOSED in the enumeration.

MartinHjelmare reviewed (approved):
Thanks!
